### ISSUE TITLE
关于训练效果以及调参方面的问题

### ISSUE THREAD

**Issue by ccuted** [bcmi/SLBR-Visible-Watermark-Removal#36]
您好！非常感谢您优秀的工作！这里我有几个问题想要请教您：
1.我在训练模型的时候发现水印预测效果很差，不知道从哪里入手去改善模型效果；
2.我发现我训练模型的时候40个epoch的效果最好，甚至比90个epoch的效果还要好,应当是过拟合了，正通过调参等操作来实现改变，请问您有比较好的建议嘛？
3.测试的时候我发现，psnr值一直没有显著提升，仅仅达到了19左右
4.我看到您在回答其他人提出的问题时涉及到了beta参数，请问在代码中具体指的是哪个参数呢？
Let's suppose that watermarked image is J, original image is I, wateramrk alpha map is A, watermark is W and controllable parameter β∈[0,1].
我在代码中有看到存在两个β参数分别是：
   parser.add_argument('--beta1', default=0.5, type=float, help='initial learning rate')
   parser.add_argument('--beta2', default=0.999, type=float, help='initial learning rate')
   以上是两个默认参数，其中的β1我改成0.9，做了实验，但是很遗憾效果并不好，所以想请问您是我理解错了嘛还是这个β参数另有？
   非常期待您的回复！感谢您的帮助！

**Comment by jimleungjing** [comment]
1. 先检查数据集的规模，SLBR模型本质也是数据驱动的，像之前有网友用100张图片训练，肯定效果不好。最起码要做到1000级别的数据。没有显卡资源做大规模训练的情况下，在我们的pre-trained model上 finetune 是最好的。其次检查数据质量，看看有没有脏数据，比如说会让模型loss 突然增大甚至NaN的数据。最后检查实验参数，一开始不熟悉调参，可以先按给出的参数跑实验
2. 水印如果是灰白色的很快就能拟合了；彩色的40epoch 性能也接近，不一定要按照90epoch 训练。从模型鲁棒性来说，还是从数据入手，让模型训练过程看到更多的不一样的数据，以及challenge 的数据
3. psnr=19，先检查MSE loss ，groud truth 有没有问题
4. beta 指的是 可控水印去除力度的参数，不是Adam optimizer 的超参数

**Comment by ccuted** [comment]
1.用的数据不算少，就是clwd的6万条数据，数据直接下载的您发布的那个clwd;其次，您说的那个预训练模型，在加载参数的时候出现了RuntimeError,显示参数结构不匹配，我无法正常使用您的预训练模型
2.40个epoch的效果还不错，但是水印还是没有完全去除，并且水印预测效果很差：
Batch[100/10017]| PSNR:19.6145 | SSIM:0.9643 | RMSE:6.9182 | RMSEw:nan | primeIoU:-0.0011, primeF1:0.5388 | maskIoU:-0.0008 | maskF1:0.5617 | time:136.32
这是测试图片的指标
3."beta 指的是 可控水印去除力度的参数，不是Adam optimizer 的超参数"
我明白您的意思，但是我不太知道在代码中具体指的是那个参数，不知道怎么修改这个参数，烦告知
能收到您的回复，不胜欣喜！万分感谢！

**Comment by ccuted** [comment]
> 1. 先检查数据集的规模，SLBR模型本质也是数据驱动的，像之前有网友用100张图片训练，肯定效果不好。最起码要做到1000级别的数据。没有显卡资源做大规模训练的情况下，在我们的pre-trained model上 finetune 是最好的。其次检查数据质量，看看有没有脏数据，比如说会让模型loss 突然增大甚至NaN的数据。最后检查实验参数，一开始不熟悉调参，可以先按给出的参数跑实验
> 2. 水印如果是灰白色的很快就能拟合了；彩色的40epoch 性能也接近，不一定要按照90epoch 训练。从模型鲁棒性来说，还是从数据入手，让模型训练过程看到更多的不一样的数据，以及challenge 的数据
> 3. psnr=19，先检查MSE loss ，groud truth 有没有问题
> 4. beta 指的是 可控水印去除力度的参数，不是Adam optimizer 的超参数

感谢您的回答！
1.我们的数据使用了您发布的clwd数据集，6万的数据规模，想来也是够的，我们正在尝试扩充数据集使用10万的数据进行训练，后续有结果，我再发上来
2.很遗憾，您发布的预训练模型我在加载时出现了runtime error,经查询发现是参数结构不匹配，还没太懂该怎样解决，如果您有好的方法提供，不胜感激！
3.默认参数下的40个epoch的训练效果算是不错了，水印很大程度上变淡了很多，但是还是没有达到论文中那样好的效果，并且水印预测效果特别差，不是论文中那样的黑白图像，是带有背景信息的灰度图，不知道是我在使用这个方法时出现了什么问题还是我训练的时候哪一步出现了错误，目前还没有找到好的办法解决，想请教您是否有好的建议提供
![1004_psnr_14 52_rmsew_39 69_f1_0 3571](https://github.com/bcmi/SLBR-Visible-Watermark-Removal/assets/115095955/3e3c6267-dae6-43ba-8d5e-9579afa159d7)

4.Batch[200/10017]| PSNR:20.0889 | SSIM:0.9602 | RMSE:7.4138 | primeIoU:-0.0014, primeF1:0.5396 | maskIoU:-0.0013 | maskF1:0.5551 | time:87.89这是默认参属下，70个epoch的测试指标，效果与40个epoch的差不多，psnr=20，还不知道从哪方面可以提高

**Comment by ccuted** [comment]
> 1. 先检查数据集的规模，SLBR模型本质也是数据驱动的，像之前有网友用100张图片训练，肯定效果不好。最起码要做到1000级别的数据。没有显卡资源做大规模训练的情况下，在我们的pre-trained model上 finetune 是最好的。其次检查数据质量，看看有没有脏数据，比如说会让模型loss 突然增大甚至NaN的数据。最后检查实验参数，一开始不熟悉调参，可以先按给出的参数跑实验
> 2. 水印如果是灰白色的很快就能拟合了；彩色的40epoch 性能也接近，不一定要按照90epoch 训练。从模型鲁棒性来说，还是从数据入手，让模型训练过程看到更多的不一样的数据，以及challenge 的数据
> 3. psnr=19，先检查MSE loss ，groud truth 有没有问题
> 4. beta 指的是 可控水印去除力度的参数，不是Adam optimizer 的超参数

5.您提到的beta参数，我在代码中没有找到对应可以修改的参数，不知可否告知？万分感谢！
期待您的回复！

**Comment by jimleungjing** [comment]
> > 1. 先检查数据集的规模，SLBR模型本质也是数据驱动的，像之前有网友用100张图片训练，肯定效果不好。最起码要做到1000级别的数据。没有显卡资源做大规模训练的情况下，在我们的pre-trained model上 finetune 是最好的。其次检查数据质量，看看有没有脏数据，比如说会让模型loss 突然增大甚至NaN的数据。最后检查实验参数，一开始不熟悉调参，可以先按给出的参数跑实验
> > 2. 水印如果是灰白色的很快就能拟合了；彩色的40epoch 性能也接近，不一定要按照90epoch 训练。从模型鲁棒性来说，还是从数据入手，让模型训练过程看到更多的不一样的数据，以及challenge 的数据
> > 3. psnr=19，先检查MSE loss ，groud truth 有没有问题
> > 4. beta 指的是 可控水印去除力度的参数，不是Adam optimizer 的超参数
> 
> 感谢您的回答！ 1.我们的数据使用了您发布的clwd数据集，6万的数据规模，想来也是够的，我们正在尝试扩充数据集使用10万的数据进行训练，后续有结果，我再发上来 2.很遗憾，您发布的预训练模型我在加载时出现了runtime error,经查询发现是参数结构不匹配，还没太懂该怎样解决，如果您有好的方法提供，不胜感激！ 3.默认参数下的40个epoch的训练效果算是不错了，水印很大程度上变淡了很多，但是还是没有达到论文中那样好的效果，并且水印预测效果特别差，不是论文中那样的黑白图像，是带有背景信息的灰度图，不知道是我在使用这个方法时出现了什么问题还是我训练的时候哪一步出现了错误，目前还没有找到好的办法解决，想请教您是否有好的建议提供 ![1004_psnr_14 52_rmsew_39 69_f1_0 3571](https://user-images.githubusercontent.com/115095955/260369994-3e3c6267-dae6-43ba-8d5e-9579afa159d7.jpg)
> 
> 4.Batch[200/10017]| PSNR:20.0889 | SSIM:0.9602 | RMSE:7.4138 | primeIoU:-0.0014, primeF1:0.5396 | maskIoU:-0.0013 | maskF1:0.5551 | time:87.89这是默认参属下，70个epoch的测试指标，效果与40个epoch的差不多，psnr=20，还不知道从哪方面可以提高

1. 预训练模型没有问题的。你按照 requirement.txt 安装环境试试。
2. 先检查训练loss，可参考我公开的pretrained model 在CLWD 数据集上的loss。然后水印数据集需要构造好，一个水印可以多次放置在不同图片上，水印种类需要多样化。最后如果想要更好的效果，可以在MSE loss 收敛后，增加GAN 的结构训练
3. psnr 才20效果太差了，公开的模型最终psnr 效果可以达到38。你的训练有问题，primeIoU, maskIoU 指标都有问题，可以检查一下训练时，预测的Mask质量
4. beta 参数，可控水印去除力度参数需要自己写代码实现，我只是提供思路。Adam optimizer beta 超参数代码初始化部分有，仔细检查吧
